verification/src/solidity/verifier.rs

```diff
@@ -1,51 +1,281 @@
+        // the whole map has been processed. That adds another layer
+        // of assurance that encoded bytes are actually metadata hash.
+        if d.datatype().is_ok() {
+            return Err(Error::custom(ParseMetadataHashError::NonExhausted));
+        }
+
+        let solc = solc.map(bytes::Bytes::copy_from_slice);
+        Ok(MetadataHash { solc })
+    }
+
+    fn nil() -> Option<Self> {
+        Some(Self { solc: None })
+    }
+}
+
 /// Wrapper under `evm.deployedBytecode` from the standard output JSON
 /// (https://docs.soliditylang.org/en/latest/using-the-compiler.html#output-description).
 ///
 /// Provides an interface to retrieve parts the deployed bytecode consists of:
-/// actual bytecode participating in EVM transaction execution and optionally metadata hash.
+/// actual bytecode participating in EVM transaction execution and metadata hash.
 #[derive(Clone, Debug, PartialEq)]
-struct DeployedBytecode {}
+struct DeployedBytecode {
+    /// Bytecode without metadata hash
+    bytecode: bytes::Bytes,
+    /// Metadata hash encoded into bytecode
+    metadata_hash: MetadataHash,
+    /// Raw deployed bytecode bytes
+    bytes: bytes::Bytes,
+}
+
+impl DeployedBytecode {
+    /// Returns deployed bytecode without metadata hash
+    pub fn bytecode(&self) -> bytes::Bytes {
+        self.bytecode.clone()
+    }
+
+    /// Returns a metadata hash
+    pub fn metadata_hash(&self) -> &MetadataHash {
+        &self.metadata_hash
+    }
+
+    /// Returns metadata hash encoded as bytes and concatenated with 2 bytes representing its length
+    pub fn encoded_metadata_hash_with_length(&self) -> bytes::Bytes {
+        let start = self.bytecode.len();
+        let end = self.bytes.len();
+        self.bytes.slice(start..end)
+    }
+}
 
 impl FromStr for DeployedBytecode {
     type Err = InitializationError;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        todo!()
+        let bytes = DisplayBytes::from_str(s)
+            .map_err(|_| InitializationError::InvalidDeployedBytecode)?
+            .0;
+
+        DeployedBytecode::try_from(bytes)
+    }
+}
+
+impl TryFrom<bytes::Bytes> for DeployedBytecode {
+    type Error = InitializationError;
+
+    fn try_from(encoded: bytes::Bytes) -> Result<Self, Self::Error> {
+        // If metadata is present, last two bytes encode its length in a two-byte big-endian encoding
+        if encoded.len() < 2 {
+            return Err(InitializationError::MetadataHashParseError(
+                "length is not encoded".to_string(),
+            ));
+        }
+
+        // Further we will cut bytes from encoded representation, but original raw
+        // bytes are still required for `DeployedBytecode.bytes`
+        // (cloning takes O(1) due to internal `bytes::Bytes` implementation)
+        let mut b = encoded.clone();
+
+        // Decode length of metadata hash representation
+        let metadata_hash_length = {
+            let b_len = b.len();
```

Comment:
?????

Reply:
does borrow checker errors out here?

Reply:
It did, but seems not to be a problem anymore. Fixed that, thanks